frontend/lib/resolution/split-init.cpp

```diff
@@ -218,27 +143,66 @@ void FindSplitInits::handleInitOrAssign(ID dstId,
     auto useType = QualifiedType(kind, rhsType.type(), p);
 
     // add the split init entry to the current frame
-    SplitInitFrame* frame = scopeStack.back().get();
-    frame->addInit(dstId, useType);
+    addInit(frame, varId, useType);
   }
 }
 
-static bool allowsSplitInit(const uast::AstNode* ast) {
-  return ast->isBlock() ||
-         ast->isConditional() ||
-         ast->isLocal() ||
-         ast->isSerial() ||
-         ast->isTry();
+void FindSplitInits::handleDeclaration(const VarLikeDecl* ast, RV& rv) {
+  addDeclaration(currentFrame(), ast);
+}
+
+void FindSplitInits::handleMention(const Identifier* ast, ID varId, RV& rv) {
+  VarFrame* frame = currentFrame();
+  SplitInitVarStatus s = findVarStatus(varId);
+
+  if (!s.foundDeclaredVar) return;
+
+  // if it wasn't already initialized or mentioned, we mark it mentioned
+  // in this frame.
+  if (!s.foundReturnsThrows && !s.foundInitedVar && !s.foundMentionedVar) {
+    frame->mentionedVars.insert(varId);
+  }
+}
+
+void FindSplitInits::handleVarAssign(const OpCall* ast, ID varId, RV& rv) {
+  // get the type for the rhs
+  auto rhsAst = ast->actual(1);
+  QualifiedType rhsType = rv.byAst(rhsAst).type();
+
+  handleInitOrAssign(varId, rhsType, rv);
+}
+
+void FindSplitInits::handleOutFormal(const FnCall* ast, const AstNode* actual,
+                                     const QualifiedType& formalType, RV& rv) {
+  ID toId = refersToId(actual, rv);
+  if (!toId.isEmpty()) {
+    handleInitOrAssign(toId, formalType, rv);
+  } else {
+    // gather any mentions used in the actual
+    handleMentions(actual, rv);
+  }
+}
+
+void FindSplitInits::handleInFormal(const FnCall* ast, const AstNode* actual,
+                                    const QualifiedType& formalType,
+                                    RV& rv) {
+  handleMentions(actual, rv);
+}
+
+void FindSplitInits::handleInoutFormal(const FnCall* ast, const AstNode* actual,
+                                       const QualifiedType& formalType,
+                                       RV& rv) {
+  handleMentions(actual, rv);
 }
 
 // updates the back frame with the combined result from
 // the then/else blocks from the conditional
-void FindSplitInits::handleExitConditional(const uast::Conditional* cond) {
-  SplitInitFrame* frame = scopeStack.back().get();
+void FindSplitInits::handleConditional(const Conditional* cond) {
+  VarFrame* frame = currentFrame();
   assert(frame->subBlocks.size() == 1 || frame->subBlocks.size() == 2);
 
-  SplitInitFrame* thenFrame = frame->subBlocks[0].frame.get();
-  SplitInitFrame* elseFrame = nullptr;
+  VarFrame* thenFrame = frame->subBlocks[0].frame.get();
```

Comment:
Since we have a `currentFrame()`, which not have a `subBlockFrame(i)` method?

Reply:
I did this one